pkg/util/wait/wait.go

```diff
@@ -237,3 +237,53 @@ func poller(interval, timeout time.Duration) WaitFunc {
 		return ch
 	})
 }
+
+type Interval struct {
+	Count    int
+	Interval time.Duration
+}
+
+func NewChannelDelayedAction(steps ...Interval) *ChannelDelayedAction {
+	return &ChannelDelayedAction{steps: steps}
+}
+
+type ChannelDelayedAction struct {
+	active <-chan time.Time
+	t      *time.Timer
+
+	steps    []Interval
+	count    int
+	interval time.Duration
+}
+
+func (q *ChannelDelayedAction) After() <-chan time.Time {
+	return q.active
+}
+
+func (q *ChannelDelayedAction) Done() {
+	q.active = nil
+}
+
+func (q *ChannelDelayedAction) Start() {
+	if q.active != nil {
+		q.count++
+		return
+	}
+	q.updateInterval()
+	q.count = 0
+	if q.t == nil {
+		q.t = time.NewTimer(q.interval)
+	} else if q.t.Reset(q.interval) {
+		<-q.t.C
+	}
+	q.active = q.t.C
+}
+
+func (q *ChannelDelayedAction) updateInterval() {
+	for i := len(q.steps) - 1; i >= 0; i-- {
```

Comment:
Adding a ton of comments

On Tue, Apr 26, 2016 at 2:50 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/util/wait/wait.go
> <https://github.com/kubernetes/kubernetes/pull/24768#discussion_r61142696>
> :
>
> > +	if q.active != nil {
> > +		q.count++
> > +		return
> > +	}
> > +	q.updateInterval()
> > +	q.count = 0
> > +	if q.t == nil {
> > +		q.t = time.NewTimer(q.interval)
> > +	} else if q.t.Reset(q.interval) {
> > +		<-q.t.C
> > +	}
> > +	q.active = q.t.C
> > +}
> > +
> > +func (q *ChannelDelayedAction) updateInterval() {
> > +	for i := len(q.steps) - 1; i >= 0; i-- {
>
> It's weird to go backwards. Also I expected the counts to be cumulative.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24768/files/91a0c160abd3ac01d0dbe9a89b3b5131c0cad419#r61142696>
>